### ISSUE TITLE
Properly wait for factories

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7174]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)| @moshez reported|
|-|-|
|Trac ID|trac#385|
|Type|defect|
|Created|2003-11-06 17:44:35Z|

<details><summary>Searchable metadata</summary>

```
trac-id__385 385
type__defect defect
reporter__moshez moshez
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__ 
keywords__ 
time__1068140675000000 1068140675000000
changetime__1469132152934470 1469132152934470
version__ 
owner__moshez moshez
cc__acapnotic cc__itamarst cc__moshez
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)|@moshez commented|
|-|-|

```
#!html
<pre>
There's a problem that TCPServer/UNIXServer don't wait for the
port to be closed down. The following patch fixes that:

Index: twisted/internet/tcp.py
===================================================================
RCS file: /cvs/Twisted/twisted/internet/tcp.py,v
retrieving revision 1.141
diff -u -r1.141 tcp.py
--- twisted/internet/tcp.py     16 Oct 2003 17:55:26 -0000      1.141
+++ twisted/internet/tcp.py     6 Nov 2003 11:43:46 -0000
@@ -656,7 +656,9 @@
         self.disconnecting = 1
         self.stopReading()
         if self.connected:
+            self.deferred = defer.Deferred()
             self.reactor.callLater(0, self.connectionLost, connDone)
+            return self.deferred

     stopListening = loseConnection

@@ -670,6 +672,7 @@
         del self.socket
         del self.fileno
         self.factory.doStop()
+        self.deferred.callback(None)

     def logPrefix(self):
         """Returns the name of my class, to prefix log entries with.
Index: twisted/application/internet.py
===================================================================
RCS file: /cvs/Twisted/twisted/application/internet.py,v
retrieving revision 1.3
diff -u -r1.3 internet.py
--- twisted/application/internet.py     24 Sep 2003 09:20:45 -0000      1.3
+++ twisted/application/internet.py     6 Nov 2003 11:43:46 -0000
@@ -59,7 +59,7 @@

     def stopService(self):
         service.Service.stopService(self)
-        self._port.stopListening()
+        return self._port.stopListening()

     def _getPort(self):
         from twisted.internet import reactor
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
Close issue 266 when this is closed.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
you'd also need to update the interface docs, write a test or two, and del the
deferred when done.

Should stopListening always return Deferred? If so any other ports should be
fixed too (e.g. UDP).
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)|@moshez commented|
|-|-|

```
#!html
<pre>
Fixed.
Fixed.
/cvs/Twisted/twisted/application/internet.py,v  &lt;--  internet.py
new revision: 1.4; previous revision: 1.3
/cvs/Twisted/twisted/internet/tcp.py,v  &lt;--  tcp.py
new revision: 1.142; previous revision: 1.141
/cvs/Twisted/twisted/test/test_tcp.py,v  &lt;--  test_tcp.py
new revision: 1.47; previous revision: 1.46
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|GitHub commented|
|-|-|

In changeset 5ed995e88fc771200923731b99e0c0b1531b8b02
```
#!CommitTicketReference repository="" revision="5ed995e88fc771200923731b99e0c0b1531b8b02"
Merge pull request #385 from einarfd/8649-einarfd-ssh-channel

Author: einarfd
Reviewer: rodrigc
Fixes: #8649 

Port twisted.conch.ssh.channel to Python 3
```
